src/main/java/com/palantir/gradle/revapi/RevapiPlugin.java

```diff
@@ -60,14 +61,40 @@ public void apply(Project project) {
                     Configuration revapiNewApi = project.getConfigurations().create("revapiNewApi", conf -> {
                         conf.extendsFrom(
                                 project.getConfigurations().getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME));
+                        conf.setCanBeConsumed(false);
                     });
 
-                    task.dependsOn(allJarTasksIncludingDependencies(project, revapiNewApi));
+                    Configuration revapiNewApiElements = project.getConfigurations()
+                            .create("revapiNewApiElements", conf -> {
+                                conf.extendsFrom(project.getConfigurations()
+                                        .getByName(JavaPlugin.API_ELEMENTS_CONFIGURATION_NAME));
+                                conf.setCanBeConsumed(false);
+                            });
+
+                    Provider<Set<Jar>> allJarTasks = allJarTasksIncludingDependencies(project, revapiNewApi);
+
+                    task.dependsOn(allJarTasks);
                     task.getAcceptedBreaks().set(acceptedBreaks(project, configManager, extension.oldGroupAndName()));
 
-                    Jar jarTask = project.getTasks().withType(Jar.class).getByName(JavaPlugin.JAR_TASK_NAME);
-                    task.getNewApiJars().set(jarTask.getOutputs().getFiles());
-                    task.getNewApiDependencyJars().set(revapiNewApi);
+                    task.getNewApiJars().set(GradleUtils.memoisedProvider(project, () -> {
+                        File thisJarFile = project.getTasks()
+                                .withType(Jar.class)
+                                .getByName(JavaPlugin.JAR_TASK_NAME)
+                                .getOutputs()
+                                .getFiles()
+                                .getSingleFile();
+
+                        Set<File> allJarFiles = allJarTasks.get().stream()
+                                .flatMap(jar -> jar.getOutputs().getFiles().getFiles().stream())
+                                .collect(Collectors.toSet());
+
+                        return Sets.union(
+                                Collections.singleton(thisJarFile),
+                                Sets.intersection(revapiNewApiElements.resolve(), allJarFiles));
+                    }));
+                    task.getNewApiDependencyJars()
+                            .set(project.provider(() -> Sets.difference(
+                                    revapiNewApi.resolve(), task.getNewApiJars().get())));
```

Comment:
I wonder if it's nicer to just work with `FileCollection` apis as they also support plus and minus, and is lazy by default so you don't need providers

Reply:
they are also `Buildable`, so the jar's `getOutputs().getFiles()` will remember what task builds it, and that will propagate to the property if you do `Property.set(that FileCollection)`

Reply:
in any case, I'd also make sure that anything you call resolve on is either added directly via `task.getInputs().files(...)`, or set on `@Classpath` properties that the task exposes